### PR TITLE
Tomcat server handlers

### PR DESCRIPTION
**Pull request by vovencij**
This is a Tomcat version of a server handler, as described and discussed in https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1886

Some notable findings:

- ~~Server instrumentation needs uncaught exceptions from outermost servlets/filters. These are passed via `AppServerBridge.THROWABLE_CONTEXT_KEY` context attribute.~~ Outermost servlets/filters set their unhandled exceptions with the current span.
- Exception handling in tomcat is done in a way, where a request passes the servlet stack two times – first pass is the initial request and the second pass is when tomcat tries to render an error page mapped in web.xml. Server span name should be updated only when doing a first pass. Thus the `AppServerBridge` context attribute and not the intuitive `callDepth` counter.

### CODE REVIEW

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      return;
+    }
+    scope.close();
+
+    if (context == null) {
+      return;
+    }
+
+    Throwable throwableToReport =
+        throwable != null ? throwable : AppServerBridge.getUnhandledThrowable(context);
```
Comment (iNikem) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:60]: Why do you need this `AppServerBridge.getUnhandledThrowable`? You can record the exception on SERVER span write when it happened inside servlet instrumentation. You don't need to pass it up.

File: smoke-tests/src/test/groovy/io/opentelemetry/smoketest/TomcatSmokeTest.groovy [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+class TomcatSmokeTest extends AppServerTest {
+
+  protected String getTargetImage(int jdk, String serverVersion) {
+    "ghcr.io/open-telemetry/java-test-containers:tomcat-${serverVersion}-jdk$jdk-20201207.405832649"
+  }
+
+  def "Server Handler test"() {
+    setup:
+    startTarget(8, "7.0.107")
```
Comment (trask) [inline smoke-tests/src/test/groovy/io/opentelemetry/smoketest/TomcatSmokeTest.groovy:19]: is this meant to be parameterized by the test params below?

File: instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    group = "org.apache.tomcat.embed"
+    module = "tomcat-embed-core"
+    // Tomcat 10 is about servlet 5.0
+    // 7.0.4 added Request.isAsync, which is needed
+    versions = "[7.0.4, 10)"
+  }
+}
+
+dependencies {
+  library group: 'org.apache.tomcat.embed', name: 'tomcat-embed-core', version: '9.0.40'
```
Comment (trask) [inline instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle:14]: typically this is the lowest version, and then tests will be run against the latest version in nightly build using -PtestLatestDeps=true, in this case you don't quite want the latest version, so need to supply `latestDepTestLibrary` like below
```suggestion
  library group: 'org.apache.tomcat.embed', name: 'tomcat-embed-core', version: '7.0.4'
  latestDepTestLibrary  group: 'org.apache.tomcat.embed', name: 'tomcat-embed-core', version: '9.+'
```

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        try {
+          servletRequest
+              .getAsyncContext()
+              .addListener(new TagSettingAsyncListener(responseHandled, context));
+        } catch (IllegalStateException e) {
+          // thrown by tomcat if request was already handled while setting the listener.
+          tracer().end(context, response);
+        }
+      }
+    }
```
Comment (trask) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:92]: Jetty and Servlet advice also does this, do you think applicable here also?
```suggestion

      // Check again in case the request finished before adding the listener.
      if (!request.isAsync() && responseHandled.compareAndSet(false, true)) {
        tracer().end(context, response);
      }
    }
```

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    Throwable throwableToReport =
+        throwable != null ? throwable : AppServerBridge.getUnhandledThrowable(context);
+
+    if (throwableToReport != null) {
+      if (response.isCommitted()) {
+        tracer().endExceptionally(context, throwableToReport, response);
+      } else {
+        tracer().endExceptionally(context, throwableToReport);
```
Comment (trask) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:66]: can you add a comment here about why not passing response when it's not committed? it's not obvious to me at least

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      @Advice.Local("otelScope") Scope scope) {
+    if (scope == null) {
+      return;
+    }
+    scope.close();
+
+    if (context == null) {
+      return;
+    }
+
```
Comment (trask) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:58]: no need for this null check I think
```suggestion
```

File: instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java [hunk 7]
```diff
@@ -32,10 +36,18 @@ public static void onEnter(
...
         scope = attachedContext.makeCurrent();
       }
 
+      // We're interested only in the very first suggested name, as this is where the initial
+      // request arrived. There are potential forward and other scenarios, where servlet path
+      // may change, but we don't want this to be reflected in the span name.
+      if (!AppServerBridge.isBetterNameSuggested(attachedContext)) {
+        tracer().updateServerSpanName(httpServletRequest);
+        AppServerBridge.setBetterNameSuggested(attachedContext, true);
+      }
```
Comment (trask) [inline instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java:49]: the SERVER span name is also updated in several of the web framework instrumentations

**Review by trask — COMMENTED** [review]
thanks @vovencij! do you think this will resolve #894?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Helper containers for the Context attributes for transferring certain information between servlet
+ * integration and app-server server handler integrations.
+ */
+public class AppServerBridge {
+
+  public static final ContextKey<AtomicReference<Throwable>> THROWABLE_CONTEXT_KEY =
```
Comment (anuraaga) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java:19]: It's better if we can have a single key / object with multiple fields for any particular concern, rather than multiple keys

File: instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java (hunk 7 above)
Comment (vovencij) [reply instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java:49]: Web Framework instrumentations are expecting the SERVER span to be present and to update the server span name unconditionally. Servlet integration was either _creating_ a new SERVER span or exiting as soon as possible if a SERVER span was already present. This flag is actually needed for the Servlet instrumentation only to indicate that the servlet instrumentation itself has set the name. I probably should rename methods to better reflect what's happening.

**Comment by vovencij** [comment]
> do you think this will resolve #894?

I think so. This is the trace I get from the demo case in that ticket in my branch:
image

And there is also an unhandled servlet exception reported in the trace:
image

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 5 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:66]: If the response is not yet committed, then response headers, including response code, are not yet written to the output stream. And the only reason response is passed to the `endExceptionally` is to read the response code.

.... And after looking at the https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/894 and into `endExceptionally` I see that in this case `http.status_code` will be set to 500 even if it wasn't actually sent. Maybe this has to be fixed also, but I'm reluctant to change that logic in the current PR.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+  public static Context init(Context ctx) {
+    return ctx.with(AppServerBridge.THROWABLE_CONTEXT_KEY, new AtomicReference<>())
+        .with(AppServerBridge.SERVLET_SUGGESTED_BETTER_NAME_KEY, new AtomicBoolean(false));
+  }
+
+  /**
+   * This is for servlet instrumentation to record exceptions unhandled by servlets and filters,
+   * that app-server integrations are interested in (to fail the span exceptionally with attached
+   * throwable), but which may be swallowed by an app-server before execution arrives to an
```
Comment (iNikem) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java:32]: Why exactly do we want to record exception in SERVER span if that was swallowed inside it?

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 5 above)
Comment (iNikem) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:66]: This comment better be in the code :)

File: instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java [hunk 10]
```diff
@@ -32,10 +36,18 @@ public static void onEnter(
 
     Context attachedContext = tracer().getServerContext(httpServletRequest);
     if (attachedContext != null) {
-      if (tracer().needsRescoping(attachedContext)) {
+      if (Servlet3HttpServerTracer.needsRescoping(attachedContext)) {
```
Comment (iNikem) [inline instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java:39]: Why this change?

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 1 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:60]: Good point! Looks that this works. Will change.

**Review by iNikem — COMMENTED** [review]
I don't like `AppServerBridge` at all. I have no better suggestions right now, but I will think about it...

File: instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java (hunk 10 above)
Comment (vovencij) [reply instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java:39]: `needsRescoping` is a static method. IntelliJ suggested.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java (hunk 9 above)
Comment (vovencij) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java:32]: I changed this part as you have suggested earlier - record throwable from the outermost servlet/filter instead of storing it in the context. Will push soon.

Comment (vovencij) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java:32]: > I don't like `AppServerBridge` at all. I have no better suggestions right now, but I will think about it...

After removing throwables from this bridge, the only thing it has left is the flag for the updated name. The flag is only used by the servlet instrumentation, but it must have the same lifetime as the context created by the app server, so this bridge must be added to the context by the server instrumentation. You are welcome to suggest a better name or  more elegant approach altogether.

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 5 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:66]: Coming soon...

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java (hunk 8 above)
Comment (vovencij) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java]: Fixed.

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+    // context must be reattached, because it has new attributes compared to the one returned from
+    // startSpan().
+    attachServerContext(context, request);
+    return context;
+  }
+
+  @Override
+  protected String getInstrumentationName() {
+    return "tomcat";
```
Comment (laurit) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java:42]: Should be "io.opentelemetry.javaagent.tomcat" to be consistent with other integrations

File: instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle (hunk 3 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle:14]: Actually I don't need `tomcat-embed-core` for the instrumentation. I only need it for the test. For the instrumentation I need two tomcat libraries - `org.apache.tomcat:tomcat-coyote` (handling of HTTP) and `org.apache.tomcat:tomcat-catalina` (Servlet implementation). However, these two dependencies must have the same version while muzzling (because some classes moved between those Jars at some point and there are `NoClassDefFound` errors from either one or other jar). Is there a way to configure muzzle in a way that two dependencies would have the same version at a time?

File: smoke-tests/src/test/groovy/io/opentelemetry/smoketest/TomcatSmokeTest.groovy (hunk 2 above)
Comment (vovencij) [reply smoke-tests/src/test/groovy/io/opentelemetry/smoketest/TomcatSmokeTest.groovy:19]: Quick answer - eventually yes, but not right now 😄 

Ideal solution eventually would be for the matrix to work as follows:

```
SmokeTestClass {
  forEach (configuration : matrix) {
    startTarget(configuration)
    testScenario1()
    testScenario2()
    ....
    tearDown(configuration)
 }
}
```

As a result there will be several separate test cases for each configuration in the matrix and the app server is started once per matrix entry.

Currently it is only possible to have either:

a) each `testScenarioX` will call `startTarget` which will effectively mean that each app server is started as many times as there are test cases. In case of this particular smoke test it would start 5 more tomcats.

or

b) There is one `testScenario` which contains all asserts in the world and it would be impossible to distinguish between failing tests.

The problem is I don't know yet how to achieve the ideal solution using neither Spock or JUnit.

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java (hunk 11 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java]: Yes, fixed.

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 6 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java]: Removed.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.ContextKey;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Helper container for Context attributes for transferring certain information between servlet
+ * integration and app-server server handler integrations.
+ */
+public class AppServerBridge {
```
Comment (iNikem) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java:16]: I now seems to me that this whole logic can live in `ServletHttpServerTracer`. This will also remove some ifs from servlet advices.

File: instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle (hunk 3 above)
Comment (trask) [reply instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle]: > Is there a way to configure muzzle in a way that two dependencies would have the same version at a time?

I don't think so. The way you have it currently using the aggregating artifact `tomcat-embed-core` seems reasonable to me.

**Review by trask — APPROVED** [review]
👍

File: instrumentation-core/servlet-2.2/src/main/java/io/opentelemetry/instrumentation/servlet/ServletHttpServerTracer.java [hunk 13]
```diff
@@ -126,4 +131,12 @@ private static String getSpanName(HttpServletRequest request) {
     }
     return spanName;
   }
+
+  public void updateServerSpanNameOnce(Context attachedContext, HttpServletRequest request) {
+    if (AppServerBridge.isPresent(attachedContext)
```
Comment (iNikem) [inline instrumentation-core/servlet-2.2/src/main/java/io/opentelemetry/instrumentation/servlet/ServletHttpServerTracer.java:136]: Is this `isPresent` needed here? The result of `if` is the same without it

File: instrumentation/servlet/servlet-2.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v2_2/Servlet2Advice.java [hunk 14]
```diff
@@ -24,14 +26,20 @@ public static void onEnter(
...
     if (!(request instanceof HttpServletRequest) || !(response instanceof HttpServletResponse)) {
       return;
     }
 
     HttpServletRequest httpServletRequest = (HttpServletRequest) request;
 
-    if (tracer().getServerContext(httpServletRequest) != null) {
+    Context serverContext = tracer().getServerContext(httpServletRequest);
+    if (serverContext != null) {
+      // We're interested only in the very first suggested name, as this is where the initial
```
Comment (iNikem) [inline instrumentation/servlet/servlet-2.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v2_2/Servlet2Advice.java:39]: Shouldn't this comment go to `updateServerSpanNameOnce` method declaration site?

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void stopSpan(
+      @Advice.Argument(0) Request request,
+      @Advice.Argument(1) Response response,
+      @Advice.Thrown Throwable throwable,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    scope.close();
```
Comment (iNikem) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:49]: `scope` can be null here

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 4 above)
Comment (iNikem) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:83]: @vovencij ?

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    implements TextMapPropagator.Getter<Request> {
+
+  private static final Logger log = LoggerFactory.getLogger(TomcatTracer.class);
+  private static final TomcatTracer TRACER = new TomcatTracer();
+
+  public static TomcatTracer tracer() {
+    return TRACER;
+  }
+
+  public Context startServerSpan(Request request, String adapterClassName) {
```
Comment (iNikem) [inline instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java:30]: Pass the whole `Method` here instead of `adapterClassName`. There is `getSpanName(Method)`

File: instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    // Tomcat 10 is about servlet 5.0
+    // 7.0.4 added Request.isAsync, which is needed
+    versions = "[7.0.4, 10)"
+  }
+}
+
+dependencies {
+  library group: 'org.apache.tomcat.embed', name: 'tomcat-embed-core', version: '7.0.4'
+  implementation project(':instrumentation:servlet:servlet-3.0:javaagent')
+  implementation("org.slf4j:slf4j-api:1.7.30")
```
Comment (iNikem) [inline instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle:16]: I wonder, why no other instrumentation defines this dependency?

File: instrumentation-core/servlet-2.2/src/main/java/io/opentelemetry/instrumentation/servlet/ServletHttpServerTracer.java (hunk 13 above)
Comment (vovencij) [reply instrumentation-core/servlet-2.2/src/main/java/io/opentelemetry/instrumentation/servlet/ServletHttpServerTracer.java:136]: `isPresent` is for the cases, where there is no app-server specific tracer (and thus no AppServerBridge in the context) which actually needs this update. Without this check camel tests, for example, will fail because the name changes unexpectedly.

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java (hunk 4 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java:83]: Checked Tomcat's source code, the check is relevant for older tomcats (7). Newer tomcats (e.g., 9) will throw IllegalStateException in this case, which is handled above. Added the check.

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java (hunk 16 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java:30]: But I don't have the method 🤔

Comment (iNikem) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java:30]: But you can have :) Like servlet advices do

File: instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle (hunk 17 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/tomcat-7.0-javaagent.gradle:16]: removed

File: instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java (hunk 16 above)
Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java:30]: Now you lost me. Servlet advices set server span name from the URL obtained from the Request. What method are you talking about?

Comment (iNikem) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java:30]: Take a look at `io.opentelemetry.javaagent.instrumentation.servlet.http.HttpServletResponseInstrumentationModule.SendAdvice`

Comment (vovencij) [reply instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/TomcatTracer.java]: ok

**Review by iNikem — APPROVED** [review]
